adoptingerlang.org

```diff
@@ -1334,6 +1334,370 @@ Repeat this exercise from time to time, and eventually do some [[https://princip
+- As part of that declarative approach, commands in Rebar3 are defined with dependency sequences. For example, the =rebar3 compile= task relies on =rebar3 get-deps= (or rather, a private form of it that locks dependencies), and will run it for you. The command =rebar3 tar= will implicitly call a sequence that includes =get-deps -> compile -> release -> tar= for you. Rebar3 therefore knows that to compile your own project, its dependencies are needed.
+- Rebar3 defines its own working area within the =_build= directory; it expects to have control over what goes in there, and for you not to track that directory in source control nor rely on its internal structure. When it builds a thing you want to use, it usually outputs the path in the terminal. For example =rebar3 escriptize= will print the path to the generated escript in =_build/default/bin/= after each run.
+- All projects are built locally within their own directory based on the Erlang runtime currently loaded in your environment
+- All dependencies are fetched to the project's =_build= directory
+- All dependencies can define their own dependencies, and rebar3 will recognize this and put them within your own project's =_build= directory as well
+- Since version numbers are unreliable (even with semantic versioning), we consider versions are information for humans, not build tools. In case of libraries that end up declared multiple times, the one closest to the project root is chosen to win ([[https://www.rebar3.org/docs/dependencies][warnings]] are emitted during the first build to let the user know about this), with the assumption that libraries declared closer to the root project are used more thoroughly (and in the case of semver issues, the root project has to pick a winner anyway, which is possible here)
+- Circular dependencies are forbidden
+- Rebar3 supports composable per-project [[https://www.rebar3.org/docs/profiles][_profiles_] that let you segment or combine configuration settings, including which dependencies are to be used only for tests, or in specific environment (say on a specific target OS). There are no restrictions on the number or names of profiles, but there are three profiles, =test=, =docs=, and =global=, which are used by rebar3 automatically for specific tasks.
+- Builds are meant to be repeatable. Dependencies are locked, and only upgraded when specifically asked to. The =rebar3.lock= file will overtake whatever versions are declared in =rebar.config= until =rebar3 upgrade= is called to change them.
+- Dependencies are always built with the =prod= profile applied, and once built will not be recompiled unless they are upgraded, even though the tool will always make sure they are there, match the lock file, and are compiled. Here "compiled" means that there is a =.app= file and for each module listed in the app file a corresponding beam file is found, but it does _not_ check if there are changes to the source file for that beam module.
```

Comment:
```suggestion
- Dependencies are always built with the =prod= profile applied, and once built will not be recompiled, but the tool will always verify they are there, match the lock file, and are compiled. Here "compiled" means that there is a =.app= file and for each module listed in the app file a corresponding beam file is found, but it does _not_ check if there are changes to the source file for that beam module. If a beam file for a module found in the =.app= file is not found it will compile the source for that beam file found in the dependency's =src=.
```

Reply:
I think that's probably too complex for just an intro of "here are the basics of how rebar3 works". It's accurate, but it's probably too much,

Reply:
Yea, but then I'd remove most of the original as well. Take it down to just that it builds in the `prod` profile and that it "verifies" the dependencies each time, which does not include scanning for modified source modules.